src/compiler/checker.ts

```diff
@@ -23925,7 +23925,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             callback(getTypeAtPosition(source, i), getTypeAtPosition(target, i));
         }
         if (targetRestType) {
-            callback(getRestTypeAtPosition(source, paramCount), targetRestType);
+            callback(getRestTypeAtPosition(source, paramCount, /*readonly*/ isConstTypeVariable(targetRestType) && !isMutableArrayOrTuple(getBaseConstraintOrType(targetRestType))), targetRestType);

```

Comment:
IIUC this won't work correctly with union constraints, for example:
```ts
declare function test<const T extends number[] | string[]>(
  f: (...args: T) => void
): T;

const res = test((a: string, b: string) => {});
```

Reply:
Yeah, this should probably use `some(...)` to distribute over unions.